src/components/implementation/tests/micro_booter/mb_tests.c

```diff
@@ -2,328 +2,153 @@
+#define debug 0
+#undef PERF
 
-static void
-thd_fn_perf(void *d)
-{
-	cos_thd_switch(BOOT_CAPTBL_SELF_INITTHD_CPU_BASE);
+unsigned int cyc_per_usec;
 
-	while (1) {
-		cos_thd_switch(BOOT_CAPTBL_SELF_INITTHD_CPU_BASE);
+int
+_expect_llu(int predicate, char *str, long long unsigned a, long long unsigned b, char *errcmp, char *testname, char * file, int line)
+{
+	if (predicate) {
+		PRINTC("%s Failure: %s @ %d: ",
+			 testname, file, line);
+		printc("(%s %lld", str, a);
+		printc(" %s %lld)\n", errcmp, b);
+		return -1;
 	}
-	PRINTC("Error, shouldn't get here!\n");
+	return 0;
 }
 
-static void
-test_thds_perf(void)
-{
-	thdcap_t  ts;
-	long long total_swt_cycles = 0;
-	long long start_swt_cycles = 0, end_swt_cycles = 0;
-	int       i;
-
-	ts = cos_thd_alloc(&booter_info, booter_info.comp_cap, thd_fn_perf, NULL);
-	assert(ts);
-	cos_thd_switch(ts);
-
-	rdtscll(start_swt_cycles);
-	for (i = 0; i < ITER; i++) {
-		cos_thd_switch(ts);
+int
+_expect_ll(int predicate, char *str, long long a, long long b, char *errcmp, char *testname, char * file, int line)
+{
+	if (predicate) {
+		PRINTC("%s Failure: %s @ %d: ",
+			 testname, file, line);
+		printc("(%s %lld", str, a);
+		printc(" %s %lld)\n", errcmp, b);
+		return -1;
 	}
-	rdtscll(end_swt_cycles);
-	total_swt_cycles = (end_swt_cycles - start_swt_cycles) / 2LL;
-
-	PRINTC("Average THD SWTCH (Total: %lld / Iterations: %lld ): %lld\n", total_swt_cycles, (long long)ITER,
-	       (total_swt_cycles / (long long)ITER));
+	return 0;
 }
 
 static void
-thd_fn(void *d)
+thd_fn_t(void *d)
 {
-	PRINTC("\tNew thread %d with argument %d, capid %ld\n",
-	       cos_thdid(), (int)d, tls_test[cos_cpuid()][(int)d]);
-	/* Test the TLS support! */
-	assert(tls_get(0) == tls_test[cos_cpuid()][(int)d]);
+	int ret = 0;
+	if (debug) PRINTC("\tNew thread %d with argument %d\n",
+	       cos_thdid(), (int)d);
+	EXPECT_LL_NEQ((int)d, 666, "Thread: Argument / Creation");
 	while (1) cos_thd_switch(BOOT_CAPTBL_SELF_INITTHD_CPU_BASE);
 	PRINTC("Error, shouldn't get here!\n");
 }
 
 static void
-test_thds(void)
+test_thds_create(void)
 {
-	thdcap_t ts[TEST_NTHDS];
-	intptr_t i;
-
-	for (i = 0; i < TEST_NTHDS; i++) {
-		ts[i] = cos_thd_alloc(&booter_info, booter_info.comp_cap, thd_fn, (void *)i);
-		assert(ts[i]);
-		tls_test[cos_cpuid()][i] = i;
-		cos_thd_mod(&booter_info, ts[i], &tls_test[cos_cpuid()][i]);
-		PRINTC("switchto %d\n", (int)ts[i]);
-		cos_thd_switch(ts[i]);
-	}
+	thdcap_t ts;
+	intptr_t i = 666;
 
-	PRINTC("test done\n");
+	ts = cos_thd_alloc(&booter_info, booter_info.comp_cap, thd_fn_t, (void *)i);
+	if (EXPECT_LL_LT(1, ts, "Thread Creation Failed"))
+		return;
+	cos_thd_switch(ts);
 }
 
-#define TEST_NPAGES (1024 * 2) /* Testing with 8MB for now */
-
 static void
-test_mem(void)
+thd_fn_mthds_ring(void *d)
 {
-	char *      p, *s, *t, *prev;
-	int         i;
-	const char *chk = "SUCCESS";
-	int fail_contiguous = 0;
+	if (count != (int) d)
+        cos_thd_switch(BOOT_CAPTBL_SELF_INITTHD_CPU_BASE);
```

Comment:
wait, perhaps "EXPECT" here too about the return value of cos_thd_switch?

Reply:
Agree 100%.

EDIT: FIXED